### PR TITLE
Apply source maps to non thrown errors.

### LINKED ISSUE #1717: Source map not applied to caught errors (and peculiarities with uncaught errors)
Example:
`caught_error.ts`:
```ts
interface A {
  a: number;
}

try {
  throw new Error("HELLO")
} catch (e) {
  console.log(e);
}
```
compiled code:
```js
try {
    throw new Error("HELLO");
}
catch (e) {
    console.log(e);
}
//# sourceMappingURL=caught_error.js.map
//# sourceURL=/Users/kevinqian/Desktop/Programming/Deno/test/caught_error.ts
```
Output:
```
Error: HELLO
    at /Users/kevinqian/Desktop/Programming/Deno/test/caught_error.ts:2:11
```
Notice the error message follows the location of compiled JS code, instead of the TS source.

In comparison,
`uncaught_error.ts`:
```ts
interface A {
  a: number;
}

throw new Error("HELLO")
```
compiled code:
```js
throw new Error("HELLO");
//# sourceMappingURL=uncaught_error.js.map
//# sourceURL=/Users/kevinqian/Desktop/Programming/Deno/test/uncaught_error.ts
```
output:
```
/Users/kevinqian/Desktop/Programming/Deno/test/uncaught_error.ts:1:6
throw new Error("HELLO");

Uncaught Error: HELLO
    at file:///Users/kevinqian/Desktop/Programming/Deno/test/uncaught_error.ts:5:7
```
Notice that the first line points to compiled JS code, yet the error string points to the source TS code.

### PR DESCRIPTION
**Pull request by kitsonk**
Fixes #1717 

This PR fixes the issue where errors are caught or created but not thrown and then logged to the console.  Currently, the stack trace would show emitted code locations and not the source location in the stack trace.  This PR ops into Rust to remap the locations based on the source maps that are part of the module meta data.

This also exposes `Deno.applySourceMap()` which takes a location and returns the original location, so custom processing of errors and error stacks can be done.

I have an integration test, but wasn't going to finalise the unit tests until I got feedback about the exposure of the `Deno.applySourceMap()`.

### CODE REVIEW

File: js/error_stack.ts [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+      .join("\n")
+  );
+}
+
+/** Sets the `prepareStackTrace` method on the Error constructor which will
+ * op into Rust to remap source code for caught errors where the `.stack` is
+ * being accessed.
+ */
+// @internal
+export function setPrepareStackTrace(ErrorConstructor: typeof Error): void {
```
Comment (ry) [inline js/error_stack.ts]: Link https://v8.dev/docs/stack-trace-api here

**Comment by kitsonk** [comment]
I am working on an addition to this though which moves the parsed source maps into the module meta data struct, as there is currently a bit of silliness in the op that creates a hash which is discarded on every op to be compatible with the existing API.

**Comment by ry** [comment]
@kitsonk I'm afraid there's been a lot of work on errors recently and that this patch may be difficult to rebase. Do you want to pursue or close and revisit later?

**Comment by kitsonk** [comment]
> Do you want to pursue or close and revisit later?

I am going to port parts over to another branch and finish it up.  It is what I will work on when I have the time.  So let's close this one for now.

**Pull request closed without merging**